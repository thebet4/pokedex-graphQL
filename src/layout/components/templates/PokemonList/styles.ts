import styled from 'styled-components/native'

export const Container = styled.View`
  flex: 1;
  margin-top: 20px;
  padding-left: 10px;
  padding-right: 10px;
  justify-content: center;
`
