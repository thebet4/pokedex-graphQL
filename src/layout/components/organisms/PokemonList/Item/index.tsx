import { Pokemon } from '@core/models/pokemon'
import React from 'react'
import { Card } from './styles'
import CardHeader from '@layout/components/molecules/PokemonCard/CardHeader'
import CardImage from '@layout/components/molecules/PokemonCard/CardImage'
import BoldText from '@layout/components/atoms/BoldText'
import PokemonTypeList from '@layout/components/molecules/PokemonCard/PokemonTypeList'

const PokemonCard = (pokemon: Pokemon) => {
  return (
    <Card type={pokemon.types[0].type.name}>
      <CardHeader pokemonId={pokemon.id} />
      <CardImage pokemonId={pokemon.id} />
      <BoldText>{pokemon.name}</BoldText>
      <PokemonTypeList types={pokemon.types} pokemonId={pokemon.id} />
    </Card>
  )
}

export default PokemonCard
