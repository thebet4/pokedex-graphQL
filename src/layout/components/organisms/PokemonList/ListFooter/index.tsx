import { colors } from '@config/theme'
import React from 'react'
import { ActivityIndicator } from 'react-native'
import { Container } from './styles'

const ListFooter = (isLoading: boolean) => {
  if (!isLoading) return null

  return (
    <Container>
      <ActivityIndicator size='large' color={colors.purple[200]} />
    </Container>
  )
}

export default ListFooter
