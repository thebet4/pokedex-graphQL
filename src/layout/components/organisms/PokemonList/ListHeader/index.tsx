import React from 'react'
import { Container, Input, SearchButtom } from './styles'
import { Feather, MaterialIcons } from '@expo/vector-icons'
import { colors } from '@config/theme'

interface IListHeaderProps {
  onSearch: (text: string) => void
}

const ListHeader = ({ onSearch }: IListHeaderProps) => {
  const [text, setText] = React.useState<string>('')

  return (
    <Container>
      <MaterialIcons name='search' size={24} color={colors.purple[100]} />

      <Input
        placeholder='Search PokÃ©mon'
        onChangeText={(text) => setText(text)}
        value={text}
      />

      <SearchButtom onPress={() => onSearch(text.toLocaleLowerCase())}>
        <Feather name='arrow-right' size={14} color={colors.base[100]} />
      </SearchButtom>
    </Container>
  )
}

export default ListHeader
