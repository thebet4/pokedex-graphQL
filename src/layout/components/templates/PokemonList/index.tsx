import { Pokemon } from '@core/models/pokemon'
import React from 'react'
import { FlatList, ListRenderItemInfo } from 'react-native'
import { Container } from './styles'
import PokemonCard from '@layout/components/organisms/PokemonList/Item'
import ListFooter from '@layout/components/organisms/PokemonList/ListFooter'
import ListHeader from '@layout/components/organisms/PokemonList/ListHeader'

interface IPokemonListProps {
  data: Pokemon[]
  isLoading?: boolean
  loadMore: () => Promise<void>
  onSearch: (text: string) => void
}

const PokemonList = ({
  data,
  isLoading = false,
  loadMore,
  onSearch,
}: IPokemonListProps) => {
  const renderItem = ({ item }: ListRenderItemInfo<Pokemon>) => {
    return <PokemonCard {...item} />
  }

  const renderHeader = () => {
    return <ListHeader onSearch={onSearch} />
  }

  return (
    <Container>
      <FlatList
        showsVerticalScrollIndicator={false}
        numColumns={2}
        data={data}
        ListHeaderComponent={renderHeader}
        renderItem={renderItem}
        keyExtractor={(item: Pokemon) => item.id}
        onEndReachedThreshold={0.05}
        onEndReached={loadMore}
        ListFooterComponent={ListFooter(isLoading)}
      />
    </Container>
  )
}

export default PokemonList
