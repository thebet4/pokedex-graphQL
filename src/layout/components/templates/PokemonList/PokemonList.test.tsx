import * as React from 'react'
import renderer from 'react-test-renderer'

import PokemonList from '@layout/components/organisms/PokemonList'

it(`renders <PokemonList /> correctly`, () => {
  const mockedData = [
    {
      id: '1',
      name: 'Bulbasaur',
      types: [
        {
          type: {
            name: 'grass',
          },
        },
      ],
      sprites: [
        {
          sprites:
            '{"front_default":"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/11.png"}',
        },
      ],
    },
  ]

  const tree = renderer
    .create(<PokemonList data={mockedData} isLoading={false} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})
