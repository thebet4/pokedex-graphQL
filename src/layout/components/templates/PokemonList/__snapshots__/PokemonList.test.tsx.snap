// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <PokemonList /> correctly 1`] = `
<RCTScrollView
  ListFooterComponent={null}
  data={
    Array [
      Object {
        "id": "1",
        "name": "Bulbasaur",
        "sprites": Array [
          Object {
            "sprites": "{\\"front_default\\":\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/11.png\\"}",
          },
        ],
        "types": Array [
          Object {
            "type": Object {
              "name": "grass",
            },
          },
        ],
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text>
        Bulbasaur
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
